### PR TITLE
controllers: use reconcilers to send webview updates

### PR DESCRIPTION
**Pull request by nicks**
Hello @milas, @landism,

Please review the following commits I made in branch nicks/uisession15:

294b972f8479aa8a7829ccd760e3d3afa7e9d8cb (2021-05-26 20:37:27 -0400)
controllers: use reconcilers to send webview updates

Code review reminders, by giving a LGTM you attest that:

* Commits are adequately tested
* Code is easy to understand and conforms to style guides
* Incomplete code is marked with TODOs
* Code is suitably instrumented with logging and metrics

### CODE REVIEW

**Comment by nicks** [comment]
This should work as a model for how to send UI components from the reconcilers down to the webview. All the reconcilers should hold a common websocket list, and pass down the updates incrementally.

**Review by milas — APPROVED** [review]
It does seem a little strange to me for the reconciler to send WS updates.

I'd almost just expect an independent type that sets its own `Watch()` on `UIResource` / `UISession` and goes from there -- very similar to store subscriber but gets notified by API watch instead of store change summary.

File: internal/hud/server/websocket.go [hunk 1]
```diff
@@ -116,8 +147,67 @@ func (ws *WebsocketSubscriber) onChangeHelper(ctx context.Context, s store.RStor
...
+		UiSession:     uiSession,
+	})
+	ws.onSessionUpdateSent(ctx, uiSession)
+}
+
+// If a session update triggered an analytics nudge, record it so that we don't
+// nudge again.
+func (ws *WebsocketSubscriber) onSessionUpdateSent(ctx context.Context, uiSession *v1alpha1.UISession) {
+	state := ws.st.RLockState()
+	surfaced := !state.AnalyticsNudgeSurfaced
```
[the remark this answers is not in the document]
Comment (nicks) [reply internal/hud/server/websocket.go:167]: nope, that wouldn't work, because you might have multiple websockets

**Comment by nicks** [comment]
re: the reconciler - ya, I think it's unfortunate. It's not really doing reconcilation. Would you feel differently if we called it something else - e.g., Notifier - but still implemented the Reconciler interface?

I understand the instinct to use a one-off Watcher here. That's what we did when we first implemented PodWatch! And that's how we learned that it's a mistake to have long-lived Watchers. Watch() is ok for short-lived tasks, but when you need to watch changes for an unbounded period of time, you need a ton of caching and reconnection logic to make it robust/performant.

**Pull request merged**